extmod/modussl_mbedtls.c

```diff
@@ -366,8 +375,40 @@ STATIC mp_uint_t socket_ioctl(mp_obj_t o_in, mp_uint_t request, uintptr_t arg, i
         mbedtls_ssl_config_free(&self->conf);
         mbedtls_ctr_drbg_free(&self->ctr_drbg);
         mbedtls_entropy_free(&self->entropy);
+    } else if (request == MP_STREAM_POLL) {
+        // If the library signaled us that it needs reading or writing, only check that direction,
+        // but save what the caller asked because we need to restore it later
+        int saved_arg = 0;
+        if (self->poll_mask && (arg & MP_STREAM_POLL_RDWR)) {
+            saved_arg = arg & MP_STREAM_POLL_RDWR;
+            arg = ~saved_arg | self->poll_mask;
+        }
+
+        // Take into account that the library might have buffered data already
+        int has_pending = 0;
+        if (arg & MP_STREAM_POLL_RD) {
+            has_pending = mbedtls_ssl_check_pending(&self->ssl);
+            if (has_pending && arg == MP_STREAM_POLL_RD) {
+                // Shortcut if we only need to read and we have buffered data, no need to go to the underlying socket
+                return MP_STREAM_POLL_RD;
+            }
+        }
+
+        // ...otherwise fall through to pass request to underlying socket
+        mp_uint_t ret = mp_get_stream(self->sock)->ioctl(self->sock, request, arg, errcode);
```

Comment:
Can you make it so there's only one of these calls to `->ioctl()` in this function?  For example, at the very end of this function do `return ret | mp_get_stream(...)->ioctl(...);`, and then make sure `ret` is correct for all paths of logic flow.

Reply:
I don't see how we could do that, given that `ret` would have to be conditional on the result of the `->ioctl()` call.

Reply:
How about something like this:
```c
    mp_uint_t ret = 0;
    if (request == MP_STREAM_CLOSE) {
        ...
    } else if (request == MP_STREAM_POLL) {
        <existing logic of this if branch up to this ioctl call>
        if (has_pending) {
            ret |= MP_STREAM_POLL_RD;
        }
    }
    ret |= mp_get_stream(self->sock)->ioctl(...);
    if (request == MP_STREAM_POLL) {
        if (ret & self->poll_mask) {
            ret |= saved_arg;
        }
    }
    return ret;
}
```
?

Note that this is purely for code size reduction, so only worth doing if it reduces compiled firmware size.

Reply:
Thank you, I adopted this in the last version.